src/sources/exec.rs

```diff
@@ -0,0 +1,886 @@
+
+impl Default for ExecConfig {
+    fn default() -> Self {
+        ExecConfig {
+            mode: Mode::Scheduled,
+            scheduled: Some(ScheduledConfig {
+                exec_interval_secs: default_exec_interval_secs(),
+            }),
+            streaming: None,
+            command: vec!["echo".to_owned(), "Hello World!".to_owned()],
+            working_directory: None,
+            include_stderr: default_include_stderr(),
+            event_per_line: default_events_per_line(),
+            maximum_buffer_size_bytes: default_maximum_buffer_size(),
+        }
+    }
+}
+
+// TODO: Maybe split into a shared codec module
+#[derive(Debug)]
+pub struct SizedBytesCodec {
+    max_length: usize,
+}
+
+impl SizedBytesCodec {
+    pub fn new_with_max_length(max_length: usize) -> Self {
+        SizedBytesCodec { max_length }
+    }
+}
+
+// TODO: If needed it might be useful to return a list of BytesMut for
+//       a single call but returning a single seems to work fine.
+impl Decoder for SizedBytesCodec {
+    type Item = BytesMut;
+    type Error = io::Error;
+
+    fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<BytesMut>, io::Error> {
+        if !buf.is_empty() {
+            let incoming_length = buf.len();
+            if incoming_length >= self.max_length {
+                // Buffer full
+                Ok(Some(buf.split_to(self.max_length)))
+            } else {
+                // Buffer not full yet
+                Ok(None)
+            }
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn decode_eof(&mut self, buf: &mut BytesMut) -> Result<Option<BytesMut>, io::Error> {
+        Ok(match self.decode(buf)? {
+            Some(frame) => Some(frame),
+            None => {
+                if !buf.is_empty() {
+                    // Send what ever is left over
+                    Some(buf.split_to(buf.len()))
+                } else {
+                    // Nothing left ot send
+                    None
+                }
+            }
+        })
+    }
+}
+
+fn default_maximum_buffer_size() -> usize {
+    // 1MB
+    1000000
+}
+
+fn default_exec_interval_secs() -> u64 {
+    60
+}
+
+fn default_respawn_interval_secs() -> u64 {
+    5
+}
+
+fn default_respawn_on_exit() -> bool {
+    true
+}
+
+fn default_include_stderr() -> bool {
+    true
+}
+
+fn default_events_per_line() -> bool {
+    true
+}
+
+fn get_hostname() -> Option<String> {
+    crate::get_hostname().ok()
+}
+
+const EXEC: &str = "exec";
+const STDOUT: &str = "stdout";
+const STDERR: &str = "stderr";
+const STREAM_KEY: &str = "stream";
+const PID_KEY: &str = "pid";
+const EXIT_STATUS_KEY: &str = "exit_status";
+const COMMAND_KEY: &str = "command";
+const EXEC_DURATION_MILLIS_KEY: &str = "exec_duration_millis";
+
+inventory::submit! {
+    SourceDescription::new::<ExecConfig>("exec")
+}
+
+impl_generate_config_from_default!(ExecConfig);
+
+impl ExecConfig {
+    pub(self) fn validate(&self) -> Result<(), ExecConfigError> {
```

Comment:
I think we more commonly leave off the visibility modifier than use `pub(self)`. They are analogous.

https://doc.rust-lang.org/reference/visibility-and-privacy.html#pubin-path-pubcrate-pubsuper-and-pubself

Reply:
Removed the visibility modifier